runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java

```diff
@@ -85,45 +123,83 @@ public void startBundle() {
     doFnRunner.startBundle();
   }
 
+  @Override
+  public void finishBundle() {
+    doFnRunner.finishBundle();
+  }
+
   @Override
   public void processElement(WindowedValue<InputT> input) {
 
     // StatefulDoFnRunner always observes windows, so we need to explode
     for (WindowedValue<InputT> value : input.explodeWindows()) {
-
       BoundedWindow window = value.getWindows().iterator().next();
-
       if (isLate(window)) {
         // The element is too late for this window.
-        droppedDueToLateness.inc();
-        WindowTracing.debug(
-            "StatefulDoFnRunner.processElement: Dropping element at {}; window:{} "
-                + "since too far behind inputWatermark:{}",
-            input.getTimestamp(),
-            window,
-            cleanupTimer.currentInputWatermarkTime());
+        reportDroppedElement(value, window);
+      } else if (requiresTimeSortedInput) {
+        processElementOrdered(window, value);
       } else {
-        cleanupTimer.setForWindow(value.getValue(), window);
-        doFnRunner.processElement(value);
+        processElementUnordered(window, value);
+      }
+    }
+  }
+
+  private void processElementUnordered(BoundedWindow window, WindowedValue<InputT> value) {
+    cleanupTimer.setForWindow(value.getValue(), window);
+    doFnRunner.processElement(value);
+  }
+
+  private void processElementOrdered(BoundedWindow window, WindowedValue<InputT> value) {
+
+    StateInternals stateInternals = stepContext.stateInternals();
+    TimerInternals timerInternals = stepContext.timerInternals();
+
+    if (!timerInternals.currentInputWatermarkTime().isAfter(value.getTimestamp())) {
+      StateNamespace namespace = StateNamespaces.window(windowCoder, window);
+      BagState<WindowedValue<InputT>> sortBuffer = stateInternals.state(namespace, sortBufferTag);
+      ValueState<Instant> minStampState = stateInternals.state(namespace, sortBufferMinStampTag);
+      sortBuffer.add(value);
+      Instant minStamp =
+          MoreObjects.firstNonNull(minStampState.read(), BoundedWindow.TIMESTAMP_MAX_VALUE);
+      if (value.getTimestamp().isBefore(minStamp)) {
+        minStamp = value.getTimestamp();
+        minStampState.write(minStamp);
+        setupFlushTimerAndWatermarkHold(namespace, minStamp);
       }
+    } else {
+      reportDroppedElement(value, window);
```

Comment:
Such internal detail needs to be more explicitly mentioned in javadoc for the annotation.

What if I am processing stream and my out of order spread can be quite large. Does it make sense for me to use this annotation or better not. Some recommendations / hints for users would be great.

This makes me hesitant to use this annotation for stream processing (not that I have use case for it right now) unless I can be sure about timing of my data stream and watermark move.
I wonder if it would make sense to give some tradeoffs between latency, buffer size, lateness and introduce a possibility to hold watermark back for some delta, smaller than max allowed lateness.

Reply:
I think that this implies that we have to take better care of allowed lateness here. Elements that are more late than allowed lateness are dropped by any stateful dofn, this is no new behavior.

Reply:
There will be no difference in dropped data when using and not using this annotation. Mailing list thread for reference: https://lists.apache.org/thread.html/c37dfb6c545fba7d794a13c507dccebb654bbd8b317dab748a6775dc%40%3Cdev.beam.apache.org%3E

Reply:
The conversation there confirms that there is on purpose different behavior of late data dropping than I expected. I will describe this in the javadoc.

Reply:
Added mention to javadoc regarding the change in late data dropping behavior.